Add `nonStrict` modifier for `record` runtypes

remnantkevin opened the pull request:
Fixes:
- #68
- #23

## Description

### Context

The default mode of a `record` runtype is **strict**, where _strictness_ refers to how a runtype treats excess keys in untrusted input data:
- **strict**: treats excess keys as an error
- **non-strict**: ignores excess keys

Because the `record` runtype is strict by default, excess keys in its input will be treated as an error:  

```ts
import * as st from "simple-runtypes"

const user = st.record({
  id: st.integer(),
  name: st.string()
})

user({ id: 1, name: "Matt" })
// => { id: 1, name: "Matt" }

user({ id: 1, name: "Matt", isAdmin: true })
// => throws st.RuntypeError: invalid keys in record

st.use(user, { id: 1, name: "Matt", isAdmin: true })
// => { ok: false, error: { reason: 'invalid keys in record [...]', ... } }
```

`simple-runtypes` currently also offers a **non-strict** version of the `record` runtype, which is called the `sloppyRecord` runtype. Because it is non-strict, `sloppyRecord` ignores excess keys:

```ts
const sloppyUser = st.sloppyRecord({
  id: st.integer(),
  name: st.string()
})

sloppyUser({ id: 1, name: "Matt" })
// => { id: 1, name: "Matt" }

sloppyUser({ id: 1, name: "Matt", isAdmin: true })
// => { id: 1, name: "Matt" }

st.use(sloppyUser, { id: 1, name: "Matt", isAdmin: true })
// => { ok: true, result: { id: 1, name: "Matt" } }
```

### `nonStrict` modifier

A non-strict modifier allows strict `record` runtypes to be changed (modified) into non-strict versions, as opposed to having a separate non-strict runtype (see `sloppyRecord` above). Some of the reasons why this is desirable are outlined in #68. One of these reasons is that it allows for separation between the record's _definition_ and the way the record is _parsed_. The modifier approach is similar to the way the `pick`, `omit`, and `partial` runtypes work (they modify a record runtype in order to create a new runtype).

> **Note**: The "sloppy" terminology has negative connotations, and it looks untidy to have the word "sloppy" across a code base. Therefore, "sloppy" has been dropped in favour of "non-strict".

`record` runtypes continue to be strict by default, and the `nonStrict` modifier is used on an existing record runtype to create a new, non-strict record runtype. For example:

```ts
const nonStrictUser = st.nonStrict(
  st.record({
    id: st.integer(),
    name: st.string()
  })
)

nonStrictUser({ id: 1, name: "Matt" })
// => { id: 1, name: "Matt" }

nonStrictUser({ id: 1, name: "Matt", isAdmin: true })
// => { id: 1, name: "Matt" }

st.use(nonStrictUser, { id: 1, name: "Matt", isAdmin: true })
// => { ok: true, result: { id: 1, name: "Matt" } }
```

> **Note**: `nonStrict` does _not_ modify the given runtype in-place, but returns a new runtype based on the given runtype.

### Notes on capabilities and limitations

#### strict and `nonStrict` can be used in the same runtypes

```ts
const user = st.record({
  id: st.integer(),
  name: st.string(),
  company: st.nonStrict(
    st.record({
      id: st.integer(),
      name: st.string()
    })
  )
})

user({ id: 1, name: "Matt", company: { id: 1, name: "EasyJet" } })
// => { id: 1, name: "Matt", company: { id: 1, name: "EasyJet" } }

user({ id: 1, name: "Matt", company: { id: 1, name: "EasyJet", code: "EJ" } })
// => { id: 1, name: "Matt", company: { id: 1, name: "EasyJet" } }

user({ id: 1, name: "Matt", company: { id: 1, name: "EasyJet" }, isAdmin: true })
// => throws st.RuntypeError: invalid keys in record
```

#### `nonStrict` is not recursive

```ts
const nonStrictUser = st.nonStrict(
  st.record({
    id: st.integer(),
    name: st.string(),
    address: st.record({
      streetNumber: st.integer(),
      streetName: st.string()
    })
  })
)

nonStrictUser({ id: 1, name: "Matt", address: { streetNumber: 42, streetName: "Sesame St" } })
// => { id: 1, name: "Matt", address: { streetNumber: 42, streetName: "Sesame St" } }

nonStrictUser({ id: 1, name: "Matt", address: { streetNumber: 42, streetName: "Sesame St" } }, isAdmin: true)
// => { id: 1, name: "Matt", address: { streetNumber: 42, streetName: "Sesame St" } }

nonStrictUser({ id: 1, name: "Matt", address: { streetNumber: 42, streetName: "Sesame St", hasAlarm: true } })
// => throws st.RuntypeError: invalid keys in record
```

#### `nonStrict` can only be applied to records

Record runtypes include: `record`, `pick`, `omit`, `partial`, `intersection` (of two records).

```ts
// ✅
// returns non-strict record runtypes 
st.nonStrict(st.record({ id: st.integer(), name: st.string() }))
st.nonStrict(st.pick(st.record({ id: st.integer(), name: st.string() }), "id"))
st.nonStrict(st.omit(st.record({ id: st.integer(), name: st.string() }), "name"))
st.nonStrict(st.partial(st.record({ id: st.integer(), name: st.string() })))
st.nonStrict(st.intersection(st.record({ name: st.string() }), st.record({ age: st.number() })))

// ❌
// throws st.RuntypeUsageError expected a record runtype
st.nonStrict(st.string())
st.nonStrict(st.array(st.string()))
st.nonStrict(st.union(st.number(), st.string()))
// ...
```

### Compatibility

`sloppyRecord` can still be used. Under the hood, a `sloppyRecord` just uses the `nonStrict` functionality.

remnantkevin commented:
@hoeck I have updated the PR description with more details and an overview of the functionality. This is based on my understanding at the moment, but any of it can change based on discussions we have.

I will add my questions and comments about the code itself and the way `nonStrict` is implemented this weekend, and hopefully we can discuss from there.

remnantkevin commented on src/nonStrict.ts:
```diff
@@ -0,0 +1,12 @@
+import { internalRecord } from './record'
+import { Runtype, RuntypeUsageError } from './runtype'
+
+export function nonStrict<T>(original: Runtype<T>): Runtype<T> {
+  const fields = (original as any).fields
+
+  if (!fields) {
+    throw new RuntypeUsageError('expected a record runtype')
+  }
+
+  return internalRecord(fields, true)
+}
```
Like pick, omit, etc., nonStrict takes in a runtype, checks that its a record runtype (i.e. has a `fields` property), and then uses `internalRecord` to create a new runtype, this time passing in `true` as `internalRecord`'s `isNonStrict` argument.

remnantkevin replied:
should `nonStrict` be in `record.ts` instead, given its connection with that runtype? or like omit, pick, etc. should it be in its own file (like it is now)?

remnantkevin replied:
- [ ] Could you though please add a short /** ... */ function definition comment to nonStrict.

remnantkevin replied:
@hoeck Added a JSDoc comment 👍 I've tried to keep it as consistent as possible with other comments, while also trying to make the description clear and concise. Let me know what you think.

I primarily use VS Code, and this is what the JSDoc comment looks like there:

<details>
<summary>Screenshot</summary>

<img width="1106" alt="image" src="https://user-images.githubusercontent.com/8270635/221395560-4ead0610-8d94-4f6f-9a33-7ed985bbfcc7.png">

</details>

hoeck replied:
Thanks, that docstring looks good and describes the nonStrict modifier very well.

remnantkevin commented on src/omit.ts:
```diff
@@ -21,6 +21,8 @@ export function omit<T, K extends keyof T>(
 
-  // TODO: keep 'sloppyness'
-  return record(newRecordFields) as Runtype<any>
+  return internalRecord(
+    newRecordFields,
+    isNonStrictRuntype(original),
+  ) as Runtype<any>
```
here and in similar runtypes: we check if the provided runtype is non-strict; if it is, the runtype returned by `omit` will also be non-strict, if not, `omit` will return the same.

`isNonStrictRuntype` just checks if the relevant symbol property is true

remnantkevin commented on src/record.ts:
```diff
@@ -27,12 +28,14 @@ function isPureTypemap(typemap: object) {
-function internalRecord(
-  typemap: { [key: string]: Runtype<any> | OptionalRuntype<any> },
-  sloppy: boolean,
+export function internalRecord(
```
I think the changes here are pretty straightforward: basically replacing sloppy with nonStrict, and adding`isNonStrict = isNonStrictRuntypeSymbol` so that we can 'pass on' the strictness when using things like `omit`, etc.

remnantkevin commented on src/runtype.ts:
```diff
@@ -183,3 +183,9 @@ export function isFail(v: unknown): v is Fail {
 
   return (v as any)[failSymbol]
 }
+
+export function isNonStrictRuntype(fn: Runtype<any>): boolean {
+  return !!(fn as any).isNonStrict
+}
+
+export const isNonStrictRuntypeSymbol = Symbol('isNonStrict')
```
Not sure if these should be here or in `src/nonStrict.ts`, or maybe somewhere else?

remnantkevin commented on test/nonStrict.test.ts:
```diff
@@ -0,0 +1,155 @@
+  st,
+} from './helpers'
+
+describe('nonStrict', () => {
```
Similar to sloppyRecord tests, except that I've focused on aspects of non-strictness

remnantkevin commented on test/nonStrict.test.ts:
```diff
@@ -0,0 +1,155 @@
+    ])
+  })
+
+  it('rejects runtypes that are not records', () => {
```
Would like to discuss whether all that is in this test is correct / how you expected `nonStrict` to work (and not work).

remnantkevin replied:
we also don't currently have a test helper that does what I'm doing in this test: that is, take in multiple runtypes and check their usage in another. the `expectXxxValues` helpers test multiple values against one runtype, where here I'm testing multiple runtypes. not sure if it's worth adding a helper (would need to check if there were other use cases)

remnantkevin commented on test/partial.test.ts:
```diff
@@ -50,4 +55,14 @@ describe('partial', () => {
     expectRejectValues(partialRt, rejectedValues)
   })
+
+  it('should preserve nonStrict', () => {
```
I've added this test here and in the pick and omit tests. Not sure how far we want to go with testing non-strictness from both sides: e.g. how `nonStrict` uses / interacts with other runtypes, and how other runtypes use / interact with `nonStrict`

hoeck replied:
Too me it looks like the right amount of testing.

Here you're checking that `omit` and `partial` are properly checking the `isNonStrict` flag/symbol and passing it when creating their `internalRecord`.

As simple-runtypes implementation is "just" functions calling functions, testing more complicated interactions is not necessary (as say opposed to a complicated runtype compiler). We know that if one function call works, nesting that will also work - that is the beauty of side-effect free functional code :grin:.

remnantkevin commented:
General question about non-strictness, records, and 'combination' runtypes (`union`, `intersection`):

Similar tp `pick`, `omit`, etc., `nonStrict` relies on the `fields` property to determine whether the provided runtype is a record runtype. The `fields` property is needed as its value is used to reconstruct a new, modified runtype from the provided runtype. However, is it the case that all the runtypes that should have `fields`, do have `fields`? Are there cases where runtypes that you would think would have `fields`, but dont, either because there is a good reason, or because of a mistake? (I think this might also relate to #89)

I haven't thought through all the possibilities, but I can take a union of records and/or a discriminated union as an example. (But `intersection` would need to be considered as well.)

```ts
// (A) union of records -> no `fields`
st.union(st.record({ name: st.string() }), st.record({ lastname: st.string() }))

// (B) discriminated union -> `unions`, but no `fields`
st.union(
  st.record({ type: st.literal("standard"), accessLevel: st.number() }),
  st.record({ type: st.literal("pro"), accessLevel: st.number(), admin: st.boolean() })
)
```

Would we expect to be able to apply `pick`, `omit`, `nonStrict`, etc. to these runtypes? Currently we can't, because both don't have `fields` property.

Does this make sense? Does it make sense to have a union of nonStrict runtypes vs apply nonStrict to a union?

To be clear, at this point, I'm not trying to say what does and does not make sense, but rather to have the conversation.

(Also consider the various `intersection` scenarios, and the case in #89)

hoeck reviewed (approved):
Wow! Great work. You've nailed all the important details (nesting, using a symbol for the metadata, ...).

Please merge it if you are happy with my comments.

Things we (I) need to do in order to release this:

* support non-strict in `intersection`s (that runtype looks quite messy to me and maybe needs a little refactoring)
* update the changelog (I need this when I update simple-runtypes at work so I know what has changed between versions)
* update the readme: remove sloppyRecord, add nonStrict documentation

Again, **HUGE** thanks for the hard and detailed work you've put into this PR!

remnantkevin commented:
Sorry for delay on this. Planning on giving it some time this weekend.

remnantkevin commented:
### TODO
- [ ] Add JSDoc comment to `nonStrict`
- [ ] Update README with documentation for `nonStrict`
- [ ] Remove `sloppyRecord` from README
- [ ] Update CHANGELOG

hoeck commented:
> General question about non-strictness, records, and 'combination' runtypes (`union`, `intersection`):
> 
> ...
> Similar to `pick`, `omit`, etc., `nonStrict` relies on the `fields` property to determine whether the provided runtype is a record runtype. The `fields` property is needed as its value is used to reconstruct a new, modified runtype from the provided runtype. However, is it the case that all the runtypes that should have `fields`, do have `fields`? Are there cases where runtypes that you would expect to have `fields` (i.e. they seem like they would be record runtypes) don't have `fields`, either because there is a good reason, or because of a mistake? (I think this might also relate to #89)
> 
> I haven't thought through all the possibilities, but I can take a union of records and/or a discriminated union as an example. (But `intersection` should be considered as well.)
> 
> ```ts
> // (A) union of records -> no `fields`
> st.union(st.record({ name: st.string() }), st.record({ lastname: st.string() }))
> 
> // (B) discriminated union -> `unions`, but no `fields`
> st.union(
>   st.record({ type: st.literal("standard"), accessLevel: st.number() }),
>   st.record({ type: st.literal("pro"), accessLevel: st.number(), admin: st.boolean() })
> )
> ```
> 

> 
> To be clear, at this point, I'm not trying to say what does and does not make sense, but rather just wanting to have the conversation.
> 
> (Also consider the various `intersection` scenarios, and the case in #89)

I think I didn't answer this question? Or did I already?

> Would we expect to be able to apply `pick`, `omit`, `nonStrict`, etc. to these runtypes? Currently we can't, because both don't have a `fields` property.
> 
> Does this make sense? Does it make sense, for example, to have a union _of_ nonStrict runtypes vs apply nonStrict _to_ a union?
You made a good point and yes, most things that work in Typescript should work in simple-runtypes (or generally in a good runtype library) too.

> Are there cases where runtypes that you would expect to have `fields` (i.e. they seem like they would be record runtypes) don't have `fields`, either because there is a good reason, or because of a mistake? (I think this might also relate to #89)

Right, not all records implement the "reflection" interface with `.fields` and other exposed metadata. That would be a next step to increase the utility of simple-runtypes :D.

#89 was made by coworker while we were translating some untyped Ruby code into Typescript and stumbled across the need to use omit on a non-record runtype.

> To be clear, at this point, I'm not trying to say what does and does not make sense, but rather just wanting to have the conversation.

Nah that's fine to suggest stuff, no need to be so defensive I am always open to discuss stuff :D and in that case it completely makes sense.

Given your specific example with unions, it can become quite tricky to exactly mimic Typescripts behavior:

https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMpjiAJnKXkDeAUMsmAJ4AOEAXMgOQDOG2uW9JyiSjjAMhABuEADZ0QAVwC2AI2hEAvkSKhIsRCgAKUAPaFOFanXqVdHUtwi8Bwsckmz5FrFNB0ZOnSIiZFywygAggg8jMgAvGgsOHjIAD7I2jrKRAD0qWQAFsBhUFYSImBhoFz2EADuyBIgwDogZFQoABSauGDAcCLIOcgARFg5YFDAMhLtwr0AlEQByAAaEYltHSIAPMGhAHzK6Vk9eYwFRd31cPZ1ALTVtfUBM43IAJqLAPKuYOshVowANAyW1iEonomyAA

So it looks like we need to keep union type info in addition to `.fields`. I you like to we can discuss this in discord or in a new issue.

remnantkevin commented:
@hoeck 

#### README

I have updated the README: removing sloppyRecord, adding nonStrict, and various other stylistic and formatting changes. I fully understand that style and formatting are subjective, so feel free to let me know if you'd prefer non/any of the changes to be reverted. Most are an attempt at making the formatting more consistent throughout, and a few are things I noticed a long the way. I changed a few things in the Usage Examples section to try to introduce concepts before they are used, and also to separate out different examples.

Other notes:
1. I can add more detail for nonStrict (e.g. make it 100% clear that it only works on records, and is not recursive). Not sure what info you'd prefer to have in the README vs in JSDoc comments. I could also provide more examples for nonStrict to make these things clear. Or, I could leave it for now and we can come back to that at a later stage.
2. I ran the `yarn build-readme`. I haven't interacted with this script before in simple-runtypes, so please let me know if anything in the README looks wrong.

As always, any feedback is appreciated. Happy to change anything.

#### CHANGELOG

Next is adding an entry to the CHANGELOG. This would be v7.2.0, right?

#### Intersections

You mentioned this previously: "support non-strict in intersections (that runtype looks quite messy to me and maybe needs a little refactoring)"

I don't feel like I have the requisite knowledge yet to be able to look into this. So, just checking: will you be looking at this?

hoeck commented:
> @hoeck
> ### README
> 
> I have updated the README: removing sloppyRecord, adding nonStrict, and various other stylistic and formatting changes. I fully understand that style and formatting are subjective, so feel free to let me know if you'd prefer non/any of the changes to be reverted. Most are an attempt at making the formatting more consistent throughout, and a few are things I noticed a long the way. I changed a few things in the Usage Examples section to try to introduce concepts before they are used, and also to separate out different examples.

:+1: that looks good thanks!

> 
> Other notes:
> 
>     1. I can add more detail for nonStrict (e.g. make it 100% clear that it only works on records, and is not recursive). Not sure what info you'd prefer to have in the README vs in JSDoc comments. I could also provide more examples for nonStrict to make these things clear. Or, I could leave it for now and we can come back to that at a later stage.

Yes, please leave that for now - in a future version nonStrict should work on any type.

>     2. I ran the `yarn build-readme`. I haven't interacted with this script before in simple-runtypes, so please let me know if anything in the README looks wrong.
Looks good to me :+1: 

> ### CHANGELOG
> 
> Next is adding an entry to the CHANGELOG. This would be v7.2.0, right?

Right, unless we decide to remove sloppyRecord in the same release then it would be 8.0.0. Or release that together with the record->object rename to not release too many major versions?

But for now you can also use "unreleased" in the changelog. I made the habit of adding all changes immediately to the log as I tend for forget what was all done once I'm ready to release a new version.

> ### Intersections
> 
> You mentioned this previously: "support non-strict in intersections (that runtype looks quite messy to me and maybe needs a little refactoring)"
> 
> I don't feel like I have the requisite knowledge yet to be able to look into this. So, just checking: will you be looking at this?

Yeah will do that thanks for the reminder!

remnantkevin commented:
> But for now you can also use "unreleased" in the changelog.

Done 👍 

> Right, unless we decide to remove sloppyRecord in the same release then it would be 8.0.0. Or release that together with the record->object rename to not release too many major versions?

Maybe?:
- 7.2.0
	- add nonStrict
	- remove sloppyRecord from docs
- 8.0.0
	- remove sloppyRecord
	- rename record to object

I understand the concern to not release too many major versions, but I also don't think it's worth bundling up too many changes into one release. Obviously up to you at the end of the day though 😄 

> Yeah will do that thanks for the reminder!

🙂

hoeck commented:
> Maybe?:
> 
>     * 7.2.0
>       
>       * add nonStrict
>       * remove sloppyRecord from docs
> 
>     * 8.0.0
>       
>       * remove sloppyRecord
>       * rename record to object
> 
> 
> I can understand the concern to not release too many major versions, but I also don't think it's worth bundling up too many changes into one release. Obviously up to you at the end of the day though smile

Sounds like a good compromise :+1:

remnantkevin commented:
🎉  

@hoeck Thanks for all the guidance and help on this 😄
